tests/returntypes/test_autograd_qnode_new.py

```diff
@@ -0,0 +1,1942 @@
+
+        hess = qml.jacobian(qml.grad(cost_fn))(x)
+
+        expected_hess = [
+            [
+                -(np.cos(b) * ((a + b) * np.cos(a) + 2 * np.sin(a))),
+                -(np.cos(b) * np.sin(a)) + (-np.cos(a) + (a + b) * np.sin(a)) * np.sin(b),
+            ],
+            [
+                -(np.cos(b) * np.sin(a)) + (-np.cos(a) + (a + b) * np.sin(a)) * np.sin(b),
+                -(np.cos(a) * ((a + b) * np.cos(b) + 2 * np.sin(b))),
+            ],
+        ]
+        assert np.allclose(hess, expected_hess, atol=10e-2, rtol=0)
+
+    def test_hessian_vector_valued_separate_args(self, dev_name, diff_method, mode, mocker, tol):
+        """Test hessian calculation of a vector valued QNode that has separate input arguments"""
+        if diff_method in {"adjoint"}:
+            pytest.skip("Test does not support adjoint.")
+
+        dev = qml.device(dev_name, wires=1)
+
+        @qnode(dev, diff_method=diff_method, interface="autograd", mode=mode, max_diff=2)
+        def circuit(a, b):
+            qml.RY(a, wires=0)
+            qml.RX(b, wires=0)
+            return qml.probs(wires=0)
+
+        a = np.array(1.0, requires_grad=True)
+        b = np.array(2.0, requires_grad=True)
+        res = circuit(a, b)
+
+        expected_res = [0.5 + 0.5 * np.cos(a) * np.cos(b), 0.5 - 0.5 * np.cos(a) * np.cos(b)]
+        assert np.allclose(res, expected_res, atol=tol, rtol=0)
+
+        jac_fn = qml.jacobian(circuit)
+        g = jac_fn(a, b)
+        assert isinstance(g, tuple) and len(g) == 2
+
+        expected_g = (
+            [-0.5 * np.sin(a) * np.cos(b), 0.5 * np.sin(a) * np.cos(b)],
+            [-0.5 * np.cos(a) * np.sin(b), 0.5 * np.cos(a) * np.sin(b)],
+        )
+        assert np.allclose(g[0], expected_g[0], atol=tol, rtol=0)
+        assert np.allclose(g[1], expected_g[1], atol=tol, rtol=0)
+
+        spy = mocker.spy(qml.gradients.param_shift, "transform_fn")
+        jac_fn_a = lambda *args: jac_fn(*args)[0]
+        jac_fn_b = lambda *args: jac_fn(*args)[1]
+        hess_a = qml.jacobian(jac_fn_a)(a, b)
+        hess_b = qml.jacobian(jac_fn_b)(a, b)
+        assert isinstance(hess_a, tuple) and len(hess_a) == 2
+        assert isinstance(hess_b, tuple) and len(hess_b) == 2
+
+        if diff_method == "backprop":
+            spy.assert_not_called()
+        elif diff_method == "parameter-shift":
+            spy.assert_called()
+
+        exp_hess_a = (
+            [-0.5 * np.cos(a) * np.cos(b), 0.5 * np.cos(a) * np.cos(b)],
+            [0.5 * np.sin(a) * np.sin(b), -0.5 * np.sin(a) * np.sin(b)],
+        )
+        exp_hess_b = (
+            [0.5 * np.sin(a) * np.sin(b), -0.5 * np.sin(a) * np.sin(b)],
+            [-0.5 * np.cos(a) * np.cos(b), 0.5 * np.cos(a) * np.cos(b)],
+        )
+        for hess, exp_hess in zip([hess_a, hess_b], [exp_hess_a, exp_hess_b]):
+            assert np.allclose(hess[0], exp_hess[0], atol=10e-2, rtol=0)
+            assert np.allclose(hess[1], exp_hess[1], atol=10e-2, rtol=0)
```

Comment:
Curious: the tolerances are relatively high, is that intentional?

Reply:
Finite diff is not performing well for second derivative, I will separate the check in the whole test file.